### PR TITLE
[clang][Darwin] Remove legacy framework search path logic in the frontend

### LINKED ISSUE #75638: [Driver][Darwin] Move the remaining bits for header path handling over to the Driver.
The remaining bits for the header path handling need to be moved over to the Driver within AddClangSystemIncludeArgs() as has been done for other OS's to date.

clang/lib/Lex/InitHeaderSearch.cpp

https://github.com/llvm/llvm-project/blob/7113c80289d62de95bca818109324601a51fefd8/clang/lib/Lex/InitHeaderSearch.cpp#L343

```
  // NOTE: some additional header search logic is handled in the driver for
  // Darwin.
  if (triple.isOSDarwin()) {
    if (HSOpts.UseStandardSystemIncludes) {
      // Add the default framework include paths on Darwin.
      if (triple.isDriverKit()) {
        AddPath("/System/DriverKit/System/Library/Frameworks", System, true);
      } else {
        AddPath("/System/Library/Frameworks", System, true);
        AddPath("/Library/Frameworks", System, true);
      }
    }
    return;
  }
```

### PR DESCRIPTION
**Pull request by ldionne**
This removes a long standing piece of technical debt. Most other platforms have moved all their header search path logic to the driver, but Darwin still had some logic for setting framework search paths present in the frontend. This patch moves that logic to the driver alongside existing logic that already handles part of these search paths.

This is intended to be a pure refactor without any functional change visible to users, since the search paths before and after should be the same, and in the same order. The change in the tests is necessary because we would previously add the DriverKit framework search path in the frontend regardless of whether we actually need to, which we now handle correctly because the driver checks for ld64-605.1+.

Fixes #75638

### CODE REVIEW

File: clang/lib/Driver/ToolChains/Darwin.cpp [hunk 1]
```diff
@@ -758,9 +758,14 @@ void darwin::Linker::ConstructJob(Compilation &C, const JobAction &JA,
...
+  // Add framework include paths and library search paths.
+  // There are two flavors:
+  // 1. The "non-standard" paths, e.g. for DriverKit:
+  //      -L<sysroot>/System/DriverKit/usr/lib
+  //      -F<sysroot>/System/DriverKit/System/Library/Frameworks
+  // 2. The "standard" paths, e.g. for macOS and iOS:
+  //      -F<sysroot>/System/Library/Frameworks
+  //      -F<sysroot>/Library/Frameworks
   {
     bool NonStandardSearchPath = false;
```
Comment (ldionne) [inline clang/lib/Driver/ToolChains/Darwin.cpp:770]: @yln Note that I think this whole block could almost go away now. I think we always do the right thing in the linker on Apple platforms, so I don't think the driver needs any special logic for `-F` and `-L` anymore, but I could be wrong. I also wanted to keep this change minimal.

Comment (yln) [reply clang/lib/Driver/ToolChains/Darwin.cpp:770]: Do you mean that `ld64-605.1` is old enough / has "aged out" by now so we can drop this special case?

Comment (ldionne) [reply clang/lib/Driver/ToolChains/Darwin.cpp:770]: Yes, that is what I suspect.

**Comment by ldionne** [comment]
@jlebar The CI seems to be failing due to a test you were the last to touch:

```
******************** TEST 'Clang :: Preprocessor/cuda-macos-includes.cu' FAILED ********************
Exit Code: 1

Command Output (stderr):
--
[...]
error: CHECK-DAG: expected string not found in input
// CHECK-DAG: ignoring nonexistent directory "/var/empty/System/Library/Frameworks"
              ^
<stdin>:1:1: note: scanning from here
```

Long story short, that test was checking for the diagnostic output from the Frontend complaining that we can't find `/var/empty/System/Library/Frameworks` as a way to ensure that it would normally try searching in those paths for frameworks. This isn't the case anymore since the frontend doesn't know anything about these search paths, and the driver handles them instead.

This shouldn't be a problem if the normal linker is being used when linking object files produced for cuda targets, since the linker on Apple platforms nowadays knows about these search paths, and Clang doesn't really need to specify them explicitly. If that's the case, the test could simply be removed. But I don't know much about cuda, so I'd rather ask in case that's not the case.

File: clang/lib/Driver/ToolChains/Darwin.cpp (hunk 1 above)
Comment (yln) [reply clang/lib/Driver/ToolChains/Darwin.cpp:770]: Got it, that would be another good cleanup!  I recommend doing this as a separate PR.

**Review by yln — APPROVED** [review]

**Comment by brad0** [comment]
Update?

**Comment by ldionne** [comment]
I think this can be merged. I would have liked to have @jlebar 's input on interactions with Cuda, but I think this is probably good enough.

**Comment by brad0** [comment]
Thank you for the effort.

**Comment by petrhosek** [comment]
This appears to have broken the LSan build which is now failing with the following error:
```
FAILED: compiler-rt/lib/lsan/CMakeFiles/clang_rt.lsan_osx_dynamic.dir/lsan_malloc_mac.cpp.o 
/Volumes/Work/s/w/ir/x/w/llvm_build/./bin/clang++ --target=x86_64-apple-darwin22.6.0 --sysroot=/Volumes/Work/s/w/ir/cache/macos_sdk/XCode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX13.3.sdk -DCOMPILER_RT_SHARED_LIB -D_DEBUG -D_GLIBCXX_ASSERTIONS -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -Dclang_rt_lsan_osx_dynamic_EXPORTS -I/Volumes/Work/s/w/ir/x/w/llvm-llvm-project/compiler-rt/lib/lsan/.. -fPIC -fvisibility-inlines-hidden -Werror=date-time -Werror=unguarded-availability-new -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wmissing-field-initializers -Wimplicit-fallthrough -Wcovered-switch-default -Wno-noexcept-type -Wnon-virtual-dtor -Wdelete-non-virtual-dtor -Wsuggest-override -Wstring-conversion -Wmisleading-indentation -Wctad-maybe-unsupported -fdiagnostics-color -ffile-prefix-map=/Volumes/Work/s/w/ir/x/w/llvm_build/runtimes/runtimes-bins=../../../llvm-llvm-project -ffile-prefix-map=/Volumes/Work/s/w/ir/x/w/llvm-llvm-project/= -no-canonical-prefixes -Wall -Wno-unused-parameter -O3 -DNDEBUG -std=c++17 -arch arm64 -arch x86_64 -isysroot /Volumes/Work/s/w/ir/cache/macos_sdk/XCode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX13.3.sdk -fPIC -stdlib=libc++ -mmacosx-version-min=10.10 -isysroot /Volumes/Work/s/w/ir/cache/macos_sdk/XCode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX13.3.sdk -fno-lto -fPIC -fno-builtin -fno-exceptions -funwind-tables -fno-stack-protector -fno-sanitize=safe-stack -fvisibility=hidden -Wthread-safety -Wthread-safety-reference -Wthread-safety-beta -O3 -g -Wno-gnu -Wno-variadic-macros -Wno-c99-extensions -nostdinc++ -fno-rtti -Wno-format -MD -MT compiler-rt/lib/lsan/CMakeFiles/clang_rt.lsan_osx_dynamic.dir/lsan_malloc_mac.cpp.o -MF compiler-rt/lib/lsan/CMakeFiles/clang_rt.lsan_osx_dynamic.dir/lsan_malloc_mac.cpp.o.d -o compiler-rt/lib/lsan/CMakeFiles/clang_rt.lsan_osx_dynamic.dir/lsan_malloc_mac.cpp.o -c /Volumes/Work/s/w/ir/x/w/llvm-llvm-project/compiler-rt/lib/lsan/lsan_malloc_mac.cpp
In file included from /Volumes/Work/s/w/ir/x/w/llvm-llvm-project/compiler-rt/lib/lsan/lsan_malloc_mac.cpp:57:
/Volumes/Work/s/w/ir/x/w/llvm-llvm-project/compiler-rt/lib/lsan/../sanitizer_common/sanitizer_malloc_mac.inc:20:10: fatal error: 'CoreFoundation/CFBase.h' file not found
   20 | #include <CoreFoundation/CFBase.h>
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~
1 error generated.
```

**Comment by aeubanks** [comment]
same in our stage 2 build

```
 [4643/5157] Building CXX object tools/dsymutil/CMakeFiles/dsymutil.dir/CFBundle.cpp.o
 FAILED: tools/dsymutil/CMakeFiles/dsymutil.dir/CFBundle.cpp.o 
 /Volumes/Work/s/w/ir/cache/builder/src/third_party/llvm-bootstrap-install/bin/clang++ -DGTEST_HAS_RTTI=0 -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -I/Volumes/Work/s/w/ir/cache/builder/src/third_party/llvm-build/Release+Asserts/tools/dsymutil -I/Volumes/Work/s/w/ir/cache/builder/src/third_party/llvm/llvm/tools/dsymutil -I/Volumes/Work/s/w/ir/cache/builder/src/third_party/llvm-build/Release+Asserts/include -I/Volumes/Work/s/w/ir/cache/builder/src/third_party/llvm/llvm/include -isystem /Volumes/Work/s/w/ir/cache/builder/src/third_party/llvm-build-tools/zstd-1.5.5/cmake_build/install/include -DSANITIZER_OVERRIDE_INTERCEPTORS -I/Volumes/Work/s/w/ir/cache/builder/src/tools/clang/scripts/sanitizers -DLIBXML_STATIC -fvisibility-inlines-hidden -Werror=date-time -Werror=unguarded-availability-new -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wmissing-field-initializers -pedantic -Wno-long-long -Wc++98-compat-extra-semi -Wimplicit-fallthrough -Wcovered-switch-default -Wno-noexcept-type -Wnon-virtual-dtor -Wdelete-non-virtual-dtor -Wsuggest-override -Wstring-conversion -Wmisleading-indentation -Wctad-maybe-unsupported -fdiagnostics-color -fprofile-instr-use="/Volumes/Work/s/w/ir/cache/builder/src/third_party/llvm-instrumented/profdata.prof" -O3 -DNDEBUG -std=c++17 -isysroot /Volumes/Work/s/w/ir/cache/osx_sdk/XCode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX13.1.sdk -mmacosx-version-min=10.12  -fno-exceptions -fno-unwind-tables -fno-asynchronous-unwind-tables -fno-rtti -MD -MT tools/dsymutil/CMakeFiles/dsymutil.dir/CFBundle.cpp.o -MF tools/dsymutil/CMakeFiles/dsymutil.dir/CFBundle.cpp.o.d -o tools/dsymutil/CMakeFiles/dsymutil.dir/CFBundle.cpp.o -c /Volumes/Work/s/w/ir/cache/builder/src/third_party/llvm/llvm/tools/dsymutil/CFBundle.cpp
 /Volumes/Work/s/w/ir/cache/builder/src/third_party/llvm/llvm/tools/dsymutil/CFBundle.cpp:15:10: fatal error: 'CoreFoundation/CoreFoundation.h' file not found
    15 | #include <CoreFoundation/CoreFoundation.h>
       |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 1 error generated.
```

if this is intended to be NFC but isn't, then we should probably revert

**Comment by aeubanks** [comment]
I've reverted this, hopefully the command line in the failing invocation I gave is enough to repro the regression, otherwise lmk what else I can do to help repro

**Comment by fhahn** [comment]
@aeubanks thanks for the revert!

**Comment by ldionne** [comment]
@aeubanks @petrhosek Can you folks share what was the top level invocations you did to get those errors? I am having trouble understanding the problem just from the pasted errors.

**Comment by aeubanks** [comment]
is the clang invocation in the logs I posted sufficient?

**Comment by ldionne** [comment]
No, I would like to reproduce locally.

**Comment by ldionne** [comment]
Ping. I don't know how to fix the issue, I need some help.

**Comment by aeubanks** [comment]
sorry, I keep missing notifications, will find a repro

**Comment by alanzhao1** [comment]
Here's what I did to repro:

`include.cc`:

```cpp
#include <CoreFoundation/CoreFoundation.h>
```

Build clang:
```
$ cmake -G Ninja -DCMAKE_BUILD_TYPE=Release -DLLVM_ENABLE_LLD=False -DLLVM_CCACHE_BUILD=True -DLLVM_ENABLE_PROJECTS="clang" -DLLVM_TARGETS_TO_BUILD="X86;ARM;AArch64" ../llvm
$ ninja clang
```

I get the following error with this patch:
```
$ bin/clang -c -o /dev/null ~/src/tests/include.cc -isysroot/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/
/Users/ayzhao/src/tests/include.cc:1:10: fatal error: 'CoreFoundation/CoreFoundation.h' file not found
    1 | #include <CoreFoundation/CoreFoundation.h>
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
1 error generated.
```


System information:

```
$ xcodebuild -version
Xcode 13.4
Build version 13F17a
$ uname -a
Darwin ayzhao-macbookpro2.roam.internal 23.3.0 Darwin Kernel Version 23.3.0: Thu Jan  4 13:52:35 PST 2024; root:xnu-10002.81.5~13/RELEASE_ARM64_T6000 arm64
```

**Comment by alanzhao1** [comment]
If we pass `-v` to Clang, we get the following include directories:

without this patch:

```
clang -cc1 version 18.0.0git based upon LLVM 18.0.0git default target arm64-apple-darwin23.3.0
ignoring nonexistent directory "/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/local/include"
ignoring nonexistent directory "/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk//Library/Frameworks"
#include "..." search starts here:
#include <...> search starts here:
 /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/c++/v1
 /Users/ayzhao/src/llvm-project/build/lib/clang/18/include
 /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include
 /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk//System/Library/Frameworks (framework directory)
End of search list.
```

with this patch:

```
clang -cc1 version 18.0.0git based upon LLVM 18.0.0git default target arm64-apple-darwin23.3.0
ignoring nonexistent directory "/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/local/include"
#include "..." search starts here:
#include <...> search starts here:
 /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/c++/v1
 /Users/ayzhao/src/llvm-project/build/lib/clang/18/include
 /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include
End of search list.
```

So it looks like this patch causes Clang to think that the directory `/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk//Library/Frameworks` exists when it doesn't.

**Comment by brad0** [comment]
Able to look at this further?

**Comment by brad0** [comment]
Ping.

**Comment by brad0** [comment]
Any chance of reviving this?

**Comment by brad0** [comment]
@ldionne Ping.

**Comment by ldionne** [comment]
@brad0 Revived as https://github.com/llvm/llvm-project/pull/120149

**Comment by ldionne** [comment]
Sorry for the spurious request for review here, this is the wrong PR.

**Pull request merged**